server/monitor-types/tls.js

```diff
@@ -0,0 +1,266 @@
+                this.log("check successful: " + message);
+                heartbeat.msg = message;
+                heartbeat.status = UP;
+            } else {
+                this.log("check failed: " + message);
+                throw new Error(message);
+            }
+        }
+    }
+
+    /**
+     * Sends the request over the given TLS socket and returns the response.
+     * @param {AbortController} aborter   Abort controller used to abort the request
+     * @param {TLSSocket}       tlsSocket TLS socket instance
+     * @param {*}               options   Monitor options
+     * @returns {Promise<string>} Server response on success or rejected promise on error
+     */
+    async getResponseFromTlsPort(aborter, tlsSocket, options) {
+        if (options.request) {
+            this.debug_log(`sending request: '${options.request}'`);
+            tlsSocket.write(options.request);
+        }
+
+        return await this.readData(aborter, tlsSocket);
+    }
+
+    /**
+     * Connects to a given host and port using native TLS or STARTTLS.
+     * @param {AbortController} aborter     Abort controller used to abort the connection
+     * @param {string}          hostname    Host to connect to
+     * @param {int}             port        TCP port to connect to
+     * @param {boolean}         useStartTls True if STARTTLS should be used, false for native TLS
+     * @returns {Promise<TLSSocket>} TLS socket instance if successful or rejected promise on error
+     */
+    async connect(aborter, hostname, port, useStartTls) {
+        if (useStartTls) {
+            const socket = new net.Socket({
+                signal: aborter
+            });
+            socket.connect(port, hostname);
+            this.debug_log("TCP connected");
+
+            await this.startTls(aborter, socket);
+            this.debug_log("STARTTLS prelude done");
+
+            const tlsSocket = await this.upgradeConnection(socket);
+            this.debug_log("TLS upgrade done");
+            tlsSocket.on("close", (hadError) => {
+                socket.end();
+            });
+            return tlsSocket;
+        } else {
+            const tlsSocket = tls.connect(port, hostname, {
+                signal: aborter,
+                servername: hostname
+            });
+            this.debug_log("TLS connected");
+            return tlsSocket;
+        }
+    }
+
+    /**
+     * Reads available data from the given socket.
+     * @param {AbortController} aborter Abort controller used to abort the read
+     * @param {*}               socket  net.Socket or tls.TLSSocket instance to use
+     * @returns {Promise<string>} Data read from the socket or rejected promise on error
+     */
+    readData(aborter, socket) {
+        return new Promise((resolve, reject) => {
+            const cleanup = function () {
+                socket.removeListener("error", onError);
+                socket.removeListener("data", onData);
+                socket.pause();
```

Comment:
could you add a comment why we pause the socket? seems a bit counterintuitive to me